### PR TITLE
boards: Add support for the HiFi4 DSP of the MIMXRT685

### PR DESCRIPTION
**Pull request by VitekST**
This PR implements a basic Zephyr port for the Cadence HiFi 4 DSP in the NXP MIMXRT685 microcontroller. Besides a barebones support it implements interrupt handling by driving the INPUTMUX peripheral, DSP code loading and run control from the ARM side of that microcontroller and basic peripherals (Flexcomm UART and GPIO) are also instantiated.

The changes are dependent on: https://github.com/zephyrproject-rtos/hal_nxp/pull/301

### CODE REVIEW

**Comment by carlescufi** [comment]
> The changes are dependent on: [zephyrproject-rtos/hal_nxp#301](https://github.com/zephyrproject-rtos/hal_nxp/pull/301)

Can you point the manifest to the `pull/301/head` revision then please?

**Comment by iuliana-prodan** [comment]
> Can you point the manifest to the `pull/301/head` revision then please?

Here's an example on how to do it: [commit 1](https://github.com/zephyrproject-rtos/zephyr/pull/62776/commits/fbdded5679a07499dfdf352369f263822546b569) from this PR: https://github.com/zephyrproject-rtos/zephyr/pull/62776/commits

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (iuliana-prodan) [inline soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:13]: Is this toolchain available in zephyr-sdk?
I'm not aware of it.

Can we use the `nxp_rt500_adsp` the one used for RT585?

Have you compiled this support? How?
Probably with xcc, right?

**Comment by dbaluta** [comment]
@VitekST I would advise going through : https://cbea.ms/git-commit/#seven-rules

With a stress on:
* User imperative mood in the subject line https://cbea.ms/git-commit/#imperative
* wrap the commit body / commit subject at max 100 chars
* add some explanation for each commit with some info for the reviewer (like WHY the commit is needed, and what brings new to the existing code).

File: drivers/gpio/gpio_mcux_lpc.c
Comment (butok) [inline drivers/gpio/gpio_mcux_lpc.c:47]: Guess, you do not want to change LPC drivers in the RTxxx commit. Do you?

Comment (VitekST) [reply drivers/gpio/gpio_mcux_lpc.c:47]: I do - I guess the driver is just named because of historical reasons. The `gpio_mcux_lpc.c` driver really is used on the RT685, despite it being an i.MX RT device and not an LPC device.

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (VitekST) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:13]: Indeed it is not.
I haven't explored compilation using the said toolchain for the RT595, as I was instructed to just focus on the port itself by my supervisor. My guess is that it could cause some problems, as the ISAs are for sure different (HiFi 4 vs. Fusion F1).

I used `xt-clang`, as `xcc` is not even present in newer versions of the Cadence toolchain.

File: drivers/misc/nxp_rtxxx_adsp_ctl/nxp_rtxxx_adsp_ctl.c
Comment (butok) [inline drivers/misc/nxp_rtxxx_adsp_ctl/nxp_rtxxx_adsp_ctl.c:65]: If to use "for" instead of "while", the code inside { } will have only one line.

File: soc/xtensa/nxp_adsp/rt6xx/include/_soc_inthandlers.h
Comment (butok) [inline soc/xtensa/nxp_adsp/rt6xx/include/_soc_inthandlers.h:1]: Copyright and License information is missed,

File: soc/xtensa/nxp_adsp/rt6xx/include/soc/memory.h
Comment (butok) [inline soc/xtensa/nxp_adsp/rt6xx/include/soc/memory.h:3]: # should be replaced by *

File: boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts
Comment (iuliana-prodan) [inline boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts]: Why do you use the Google copyright?

It should be NXP, or if you used the rt5xx as example, you should add NXP copyright also.

File: soc/xtensa/nxp_adsp/Kconfig.defconfig
Comment (iuliana-prodan) [inline soc/xtensa/nxp_adsp/Kconfig.defconfig:3]: s/Copyright (c) 2021,2023 NXP/Copyright 2021, 2023 NXP

File: boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts
Comment (VitekST) [reply boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts]: That was done in a following commit.

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (iuliana-prodan) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:13]: > Indeed it is not. I haven't explored compilation using the said toolchain for the RT595, as I was instructed to just focus on the port itself by my supervisor. My guess is that it could cause some problems, as the ISAs are for sure different (HiFi 4 vs. Fusion F1).
> 
Oh, you're right.

> I used `xt-clang`, as `xcc` is not even present in newer versions of the Cadence toolchain.

Most probably will need a toolchain for this target in zephyr sdk.
See examples: https://github.com/zephyrproject-rtos/sdk-ng/commit/7d1b95ecfa5b9ec45319eb4dda27dd6ce9bb94dc

Comment (dbaluta) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:13]: @VitekST  we will definitely need toolchain support in sdk-ng for your platform. As @iuliana-prodan  pointed the process can be easily done following an older platform addition. 

Mind that it will take some time until you merge the toolchain support and then we update the CI to use the new SDK. Expect 2-3 weeks on that if you follow up with maintainers once you submit your changes. 

We can help you with that, but you should just look at how we enabled rt595 or imx8ulp.

Without the toolchain support in the sdg-ng we cannot merge this PR as we need CI to compile it first.

File: soc/xtensa/nxp_adsp/rt6xx/include/soc/memory.h
Comment (dbaluta) [reply soc/xtensa/nxp_adsp/rt6xx/include/soc/memory.h]: Official NXP Copyright is : ```Copyright 2023 NXP```

File: boards/xtensa/nxp_adsp_rt685/Kconfig
Comment (dcpleung) [inline boards/xtensa/nxp_adsp_rt685/Kconfig:2]: I think it is better to add to copyright when the files are first introduced, or first modified in the PR.

File: soc/xtensa/nxp_adsp/CMakeLists.txt
Comment (dcpleung) [inline soc/xtensa/nxp_adsp/CMakeLists.txt:3]: This kind of changes should be squashed back to the commit introducing these lines originally.

File: include/zephyr/dt-bindings/misc/nxp-inputmux-rt6xx.h
Comment (dcpleung) [inline include/zephyr/dt-bindings/misc/nxp-inputmux-rt6xx.h]: Could you find a way to unify this block of changes with previous commits? This is being introduced a few commits back but then are removed here, all in the same PR.

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (danieldegrasse) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:13]: Just to be clear- since the imx8ulp has the same HiFi4 DSP, can we reuse the existing toolchain? Or will we need a new toolchain for this HiFi4 specifically?

Comment (VitekST) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:13]: @danieldegrasse I haven't studied the differences between the two instances of the DSP, but from what I've encountered, they can differ, as the Xtensa IP is a customizable one. Luckily, @iuliana-prodan told me over internal Teams that adapting GCC with related tools and libraries should be just a matter of collecting a few source files from Cadence's BSP, creating a new crosstool-ng overlay with it and compiling the toolchain.

**Review by danieldegrasse — COMMENTED** [review]
Thanks for this work. A general comment about commit formatting, somewhat echoing what @dbaluta said:

Commits should follow the following rules:
- focus on one area (SOC enablement, driver changes/new driver, new board, etc)
- each commit should build correctly
- fixup commits should be avoided. Squash changes in response to reviewers back into the original commit

From a broad view, I might suggest the following set of commits:
- changes to LPC GPIO driver for RT685 DSP
- new inputmux driver
- SOC level definition for RT685 DSP
- RT685 DSP devicetree
- RT685 DSP board definition

File: boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts
Comment (iuliana-prodan) [reply boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts]: yes, don't add a new commit to fix something from a commit that was not merged.

File: include/zephyr/drivers/misc/nxp_rtxxx_adsp_ctl/nxp_rtxxx_adsp_ctl.h
Comment (VitekST) [reply include/zephyr/drivers/misc/nxp_rtxxx_adsp_ctl/nxp_rtxxx_adsp_ctl.h]: It essentially can.

File: include/zephyr/dt-bindings/misc/nxp-inputmux-rt6xx.h
Comment (VitekST) [reply include/zephyr/dt-bindings/misc/nxp-inputmux-rt6xx.h]: I redid the whole commit history, so no commits without comments and subsequent commits are there.

File: boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts
Comment (VitekST) [reply boards/xtensa/nxp_adsp_rt685/nxp_adsp_rt685.dts]: Commit history redone - no fixup commits.

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (danieldegrasse) [inline soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:45]: This line should be removed from this commit

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (danieldegrasse) [inline soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:48]: See comment above about `default y` in the `nxp_inputmux/Kconfig` file- if `default y` is present there then these lines won't be needed.

File: dts/xtensa/nxp/nxp_rt685.dtsi
Comment (danieldegrasse) [inline dts/xtensa/nxp/nxp_rt685.dtsi:1]: I'm a bit unclear why this entire file has changed in this commit- it looks like maybe spacing was changed? Either way, this change should not be in this commit. Rather, any spacing fixes needed should be squashed into the original commit making the change that needed the spacing fixed.

**Review by danieldegrasse — COMMENTED** [review]
PR is looking a lot better now, thanks for the cleanups :). One thing- some of your commits use language like "This commit does xxx". I would format them closer to how Linux requests you word commits (https://docs.kernel.org/process/submitting-patches.html)

> Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behaviour.

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.soc
Comment (VitekST) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.soc:1]: Yes. The file was copied from the RT595 Fusion F1 PR.

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.series
Comment (VitekST) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.series:1]: @danieldegrasse Same argument as above - file was copied from the RT595 Fusion F1 PR.

File: drivers/misc/nxp_inputmux/Kconfig
Comment (VitekST) [reply drivers/misc/nxp_inputmux/Kconfig:6]: @danieldegrasse This is set in `soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series`. Should I move it here?

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.soc
Comment (danieldegrasse) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.soc:1]: Understood- I'll go ahead and resolve the similar comments to these then :)

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (VitekST) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:30]: @danieldegrasse It does not.

File: soc/xtensa/nxp_adsp/rt6xx/include/_soc_inthandlers.h
Comment (kartben) [reply soc/xtensa/nxp_adsp/rt6xx/include/_soc_inthandlers.h:2]: FWIW I think the "(c)" is actually pretty common

```
❯ git grep "Copyright [^(]" | wc -l
1861
❯ git grep "Copyright (c)" | wc -l
20743
❯ git grep "Copyright [^(].*NXP"  | wc -l
729
❯ git grep "Copyright (.*NXP"  | wc -l
569
```

File: boards/arm/mimxrt685_evk/doc/index.rst
Comment (kartben) [inline boards/arm/mimxrt685_evk/doc/index.rst:344]: use ``.. zephyr-app-commands::``

**Review by kartben — DISMISSED** [review]

File: boards/arm/mimxrt685_evk/doc/index.rst
Comment (VitekST) [reply boards/arm/mimxrt685_evk/doc/index.rst]: @kartben Change committed locally and force pushed in a modified originating commit rather than committed here.

File: soc/xtensa/nxp_adsp/common/include/soc.h
Comment (VitekST) [reply soc/xtensa/nxp_adsp/common/include/soc.h]: @danieldegrasse Okay, code moved.

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (VitekST) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series]: @dbaluta @iuliana-prodan Toolchain PR created: https://github.com/zephyrproject-rtos/sdk-ng/pull/721

**Review by andyross — COMMENTED** [review]
Just discovered this in my queue.

Can you clarify the naming here?  "HiFi4" is a coprocessor defined in the Xtensa architecture.  I saw this PR and immediately got excited because I thought it meant "Someone is submitting arch-layer integration for the SIMD instructions".

But in fact this is just a SOC/Board integration for the DSP component of this bigger Cortex-M33 device.   And it happens to be a Xtensa core with HiFi4 instantiated.

I'd just call it "DSP" or "Xtensa CPU" or something?   Note that Zephyr doesn't have proper coprocessor support yet (though there is an integration for the scalar FPU which shares some properties), so you can't actually use the HiFi4 instructions anyway.

**Comment by andyross** [comment]
Agreed marketing gets confusing.  All I was really saying was that a PR named "add support or the HiFi4 DSP" sounds like you're adding support for SIMD register state and not just a board. :)

It looks like the naming in the filesystem is "nxp_adsp" though, which is perfectly clear.   No -1 from me.

**Comment by dbaluta** [comment]
> Agreed marketing gets confusing. All I was really saying was that a PR named "add support or the HiFi4 DSP" sounds like you're adding support for SIMD register state and not just a board. :)

Agree, coming from @andyross  direction it is a little bit confusing. But this is our like our 5th IMX platform which instantiates HIFI4 DSP and we used this formulation.

We could eventually say: "Add support for enabling HIFI4 DSP instance on MIMXRT685"

File: drivers/misc/nxp_rtxxx_adsp_ctl/Kconfig
Comment (danieldegrasse) [inline drivers/misc/nxp_rtxxx_adsp_ctl/Kconfig:6]: Is the lack of a `default y` here intentional? Usually we add one so all the use needs to do is enable the devicetree node, but I understand if we want to avoid `default y` here since the driver is in the MISC folder (and therefore does not depend on a subystem specific Kconfig)

File: dts/bindings/misc/nxp,rtxxx-adsp-ctl.yaml
Comment (danieldegrasse) [inline dts/bindings/misc/nxp,rtxxx-adsp-ctl.yaml:17]: Why does the comment here state "most likely clkctl0" ?

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (danieldegrasse) [inline soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:29]: Are the `default n` values needed here? IE if we drop them, would these symbols be enabled?

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.soc
Comment (danieldegrasse) [inline soc/xtensa/nxp_adsp/rt6xx/Kconfig.soc:12]: We don't need to select these Kconfigs to enable the drivers, that is now driven by devicetree. SOCs that select this currently just do so for legacy reasons (and can be cleaned up)

File: soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series
Comment (VitekST) [reply soc/xtensa/nxp_adsp/rt6xx/Kconfig.defconfig.series:29]: @danieldegrasse Yes, they are needed and they will be enabled. It seems like only `NXP_ADSP_SIGN_IMAGE` is redundant, I'll remove that.

File: drivers/misc/nxp_rtxxx_adsp_ctl/Kconfig
Comment (VitekST) [reply drivers/misc/nxp_rtxxx_adsp_ctl/Kconfig:6]: @danieldegrasse It's not actually. It compiles regardless of its inclusion, but I'll add it in.

File: dts/bindings/misc/nxp,rtxxx-adsp-ctl.yaml
Comment (VitekST) [reply dts/bindings/misc/nxp,rtxxx-adsp-ctl.yaml:17]: @danieldegrasse That is an artifact of a thought process that was wrong. Will remove.

File: drivers/gpio/gpio_mcux_lpc.c
Comment (dbaluta) [inline drivers/gpio/gpio_mcux_lpc.c:427]: @VitekST Lets put this in another PR so that we can merge it faster. 

As per discussion with @stephanosio  the toolchain will be available at the beginning of February.

https://github.com/zephyrproject-rtos/sdk-ng/pull/721#issuecomment-1886648261

Comment (VitekST) [reply drivers/gpio/gpio_mcux_lpc.c:427]: @dbaluta Okay, I'll create one.

Comment (VitekST) [reply drivers/gpio/gpio_mcux_lpc.c:427]: @dbaluta Here you go: https://github.com/zephyrproject-rtos/zephyr/pull/67667

File: boards/arm/mimxrt685_evk/doc/index.rst
Comment (danieldegrasse) [inline boards/arm/mimxrt685_evk/doc/index.rst:68]: Documentation here needs to be fixed- there is no M4 core, the RT685 has an CM33 and HiFi4 DSP

Comment (VitekST) [reply boards/arm/mimxrt685_evk/doc/index.rst]: I for some reason thought the thing had a Cortex-M3 core. Nevermind, fixed.

File: west.yml
Comment (dbaluta) [inline west.yml:253]: Please update this with the hash of the merged commit or last existing commit in hal_xtensa.

**Review by nordicjm — CHANGES_REQUESTED** [review]
Needs to be migrated to hwmv2

**Comment by dbaluta** [comment]
@VitekST do you have bandwith to look at porting this to hardware model v2?

You should basically follow the similar change here: https://github.com/zephyrproject-rtos/zephyr/pull/69269

**Comment by VitekST** [comment]
@dbaluta I don't think I have that much time to spare until I have to turn in my thesis about porting Zephyr and SOF to the target platform, so I would rather focus on the SOF side of things. At the very least, I don't see the point of making the changes until the `collab-hwm` branch is merged.

**Comment by dbaluta** [comment]
@VitekST  ok, no problem. I wil keep an eye on this and see if can convert the patches next week.

File: soc/xtensa/nxp_adsp/rt6xx/include/soc/memory.h
Comment (hakehuang) [inline soc/xtensa/nxp_adsp/rt6xx/include/soc/memory.h:7]: need change to ifndef ...

**Comment by dbaluta** [comment]
@VitekST  started on porting this to hwmv2. First patch of the series with some preparation is here: https://github.com/zephyrproject-rtos/zephyr/pull/70952

Will close this PR and send a follow up with support for hifi4.

**Comment by VitekST** [comment]
@dbaluta Lovely, thank you.

**Pull request closed without merging**
